### ISSUE TITLE
Push/Pull To/from Integration Fails with SugarCRM Plugin Version 5+

### ISSUE THREAD

**Issue by pstevens71** [mautic/mautic#13796]
### Mautic Version

5.0.x series

### Way of installing

I downloaded a release from https://www.mautic.org/mautic-releases

### PHP version

8.1

### What browsers are you seeing the problem on?

Chrome

### What happened?

I've been using the SugarCRM integration for SuiteCRM for years.  It has worked pretty well, with minor flaws.  Recently I've noticed SuiteCRM-> Mautic works fine.  New records are created on both sides.  However, it does not push back to SuiteCRM where Mautic field is the primary OR the SuiteCRM field is blank and Mautic has a value (like it used to).  I'm getting the following error.  This one is a little too complicated for me to troubleshoot, hoping for some help:   

mautic.ERROR: INTEGRATION ERROR: Sugarcrm - Undefined method "detachEntities". The method name must start with either findBy, findOneBy or countBy! [] {"hostname":"xxx.xxx.xx.xxx.host.secureserver.net","pid":3}

UPDATE:  only one record syncs and the then the sync stops. Even if you synccontacts -- "10 years" or -- fetch all, in Version 5 no sync happens.  If I run the command -vvv I can see that it successfuly connects, gets the list of users, then ends.  No records are added (the first run produces one contact and one lead, none after that)

I've tested up to 4.4.11 and it works, but not after.

### How can we reproduce this issue?

Step 1:  Sync Matuic with SuiteCRM make changes on either side, the error is generated when the synccontacts job runs.

### Relevant log output

```shell
mautic.ERROR: INTEGRATION ERROR: Sugarcrm - Undefined method "detachEntities". The method name must start with either findBy, findOneBy or countBy! [] {"hostname":"xxx.xxx.xx.xxx.host.secureserver.net","pid":3}

Also:

mautic.ERROR: INTEGRATION ERROR: Sugarcrm - cURL error 28: Resolving timed out after 10001 milliseconds (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) for https://xxxxxxxxxxxxxxxxxx.com/service/v4_1/rest.php [] {"hostname":"xxxx.xxxx.xxxx.host.secureserver.net","pid":3}
```

### Code of Conduct

- [X] I confirm that I have read and agree to follow this project's Code of Conduct

<br /><hr>
Care about this issue? Want to get it resolved sooner? If you are a <a href='https://www.mautic.org/become-a-member-of-mautic'>member of Mautic</a>, you can add some funds to the <a href='https://opencollective.com/mautic/projects/bounties'>Bounties Project</a> so that the person who completes this task can claim those funds once it is merged by a member of the core team! Read the docs <a href='https://contribute.mautic.org/product-team/mautic-bounty-programme'>here.</a>

**Comment by RCheesley** [comment]
Thanks for reporting the bug @pstevens71 and sorry for the delay in triaging - hopefully we can get someone to take a look at this soon!

**Comment by pstevens71** [comment]
Thanks @RCheesley I really need some help with this one.  I tried a bunch of stuff to resolve the error, but it seems this level of PHP is a little outside my scope!  Hopefully, a more seasoned dev will chime in, and I'm happy to do the leg work and testing with a little guidance!

**Comment by pstevens71** [comment]
If it helps I've done a whole bunch of testing.  4.4.11 works fine.  I couldn't try 4.4.12 as I was running into DB install issues.  I have an inistall that's been working for years, that I upgraded to version 5, that still syncs  (I really think the issue comes down to database issues).  I had a number of db issues trying to install version 5.

Here are some of them:
(2 files changed)
app/migrations/Version20210112162046.php
app/bundles/IntegrationsBundle/Entity/ObjectMapping.php

integration_integration_object_name_last_sync_date

(one file changed)
mautic/app/bundles/IntegrationsBundle/Entity/FieldChange.php

integration_object_type_modification_composite_key

In both cases the key name is too long and MySQL doesn't support that length of key name.  So I shortened them to be able to install it.  (I was thinking this wouldn't have any ill affects as it's just the key).

I'm also getting this error about "json_array" that I haven't been able to solve.

Unknown column type "json_array" requested. Any Doctrine type that you use has to be registered with \Doctrine\DBAL\Types\Type::addType(). You can get a list of all the
  known types with \Doctrine\DBAL\Types\Type::getTypesMap(). If this error occurs during database introspection then you might have forgotten to register all database type
  s for a Doctrine Type. Use AbstractPlatform#registerDoctrineTypeMapping() or have your custom types implement Type#getMappedDatabaseTypes(). If the type name is empty yo
  u might have a problem with the cache or forgot some mapping information.


Basically what happens is if run the synccontacts command in terminal with -vvv I can see that it connects OK, then it get's the list of users OK, and typically, I gets ONE contact and ONE lead.  And then it fails looping through them all.

I keep getting:  [2024-08-02T01:30:41.429440+00:00] mautic.ERROR: INTEGRATION ERROR: Sugarcrm - Undefined method "detachEntities". The method name must start with either findBy, findOneBy or countBy! [] {"hostname":"xxxxxxxxxxxxxxxxx.host.secureserver.net","pid":14510}

**Comment by RCheesley** [comment]
I've labelled this as a regression as it seems like we've broken something between releases.

**Comment by RCheesley** [comment]
@pstevens71 I found mention of the json_array error in this issue: https://github.com/mautic/mautic/issues/13281 - maybe some of the steps there might help with digging into that one, at least?

**Comment by pstevens71** [comment]
Thanks @RCheesley re the #13281 if memory serves me correctly there was a docterine file that added the json array and another that removed the json array.  So I just did a grep to search for it in the docterine files and removed it from both files (the addtion and the removal) and this allowed me to install./upgrade.

**Comment by pstevens71** [comment]
Update!  I haven't tested this yet, but I don't want this to close due to inactivity!  One of my clients has reported that in  mautic/plugins/MauticCrmBundle/Integration/SugarcrmIntegration.php, simply commenting out the line:
913   :                  $this->integrationEntityModel->getRepository()->detachEntities($integrationEntities);
And replacing  it with the line from version 4:  $this->em->clear('Mautic\PluginBundle\Entity\IntegrationEntity');

Solve the problem and the sync works again.  This PHP is a little over my head, so I don't know what else this might effect.  I will test in my DEV and report back.

I see this detachEntities is being used in all the CRM integration plugins.  Just wondering if all the integrations are a problem or is this unique to SuiteCRM?

**Comment by shubham-pawar** [comment]
At least found the 4 files with same code: 

[Code files](https://github.com/search?q=repo%3Amautic%2Fmautic+%24this-%3EintegrationEntityModel-%3EgetRepository%28%29-%3EdetachEntities%28%24integrationEntities%29%3B&type=code)

**Comment by pstevens71** [comment]
@shubham-pawar Exactly!  That's why I'm wondering if other integrations are affected by this.

**Comment by pstevens71** [comment]
I've tested this and it does work!!!!!  Changing the line back to what it was in version 4.11 works!
However, new problem in version 5.  When "update emails do not contact" is selected, call to undefined member function getDoNotContact() on null.  I'll have to open another ticket for that.

**Comment by pstevens71** [comment]
UPDATE:  I've done further testing and everything appears to work fine.  The issue I was having will the "do not contact" was that I had a couple of test records in my DEV I created by API for another project and the "do not contact" in suitecrm was not populated and thus "null" and the sync crashed out because it couldn't deal with the null.  I removed these two records and it works fine.

Hey @RCheesley I would appreciate if a dev that was involvded in the change for version 5 would review the effect of changing that one line back to what it was in Ver. 4.  I'm not really sure of all the implications, other that it works now.  If this seems like a good fix, I'll be happy to enter a PR for the change.

**Comment by mautibot** [comment]
This issue has been mentioned on **Mautic Forums**. There might be relevant details there:

https://forum.mautic.org/t/suitecrm-mautic-actual-situation/13909/24

**Comment by RCheesley** [comment]
@mautic/core-team would be great to get some 👀 on [this comment](https://github.com/mautic/mautic/issues/13796#issuecomment-2346259546).

**Comment by Creativologist** [comment]
Per  [pstevens71]([url] https://github.com/pstevens71(url))  fix

Getting this in Mautic 6.06

In SugarcrmIntegration.php line 907:
                                                                                       
  Undefined constant "MauticPlugin\MauticCrmBundle\Integration\‘Mautic\PluginBundle\E  
  ntity\IntegrationEntity’" 

That line is
$this->em->clear('Mautic\PluginBundle\Entity\IntegrationEntity');

**Comment by shubham-pawar** [comment]
## Changes to make

### File location:  

> mautic/plugins/MauticCrmBundle/Integration/SugarcrmIntegration.php
- Around code lines 906 and 1621

### Remove the below line:
```
$this->integrationEntityModel->getRepository()->detachEntities($integrationEntities);
```

### Replace with the below line:
```
$this->em->clear('Mautic\PluginBundle\Entity\IntegrationEntity');
```
